### PR TITLE
Experimental code for shared memory reservations

### PR DESCRIPTION
**Pull request by Nekotekina**
If there is no noticeable degradation, I may consider using it. Currently it's very unclear.
Both TSX and non-TSX CPUs are welcomed to test.

### CODE REVIEW

File: rpcs3/Emu/Memory/vm.cpp [hunk 1]
```diff
@@ -119,10 +119,38 @@ namespace vm
 		_register_lock(&cpu);
 	}
 
-	atomic_t<u64>* passive_lock(const u32 addr, const u32 end)
+	atomic_t<u64>* passive_lock(u32 addr /*mutable*/, u32 end /*mutable*/)
 	{
-		static const auto test_addr = [](const u32 target, const u32 addr, const u32 end)
+		if (addr < 0x10000 + g_locations[vm::main]->size || addr >= 0xc0000000)
```
Comment (elad335) [inline rpcs3/Emu/Memory/vm.cpp:124]: stack memory allows reservations.

Comment (Nekotekina) [reply rpcs3/Emu/Memory/vm.cpp:124]: Well, I didn't assume otherwise. I only assumed it didn't allow oddities (shared memory).

**Comment by legend800** [comment]
Didn't see any regressions in heavy games from my testing like GOW, Wipeout, MS Apocalypse.

GOW3 PR:
2020-04-11 08_55_27-FPS_ 17 05 _ Vulkan _ 0 0 9-156135ba-rsrv _ God of War® III  BCUS98111

Master:
2020-04-11 09_00_17-FPS_ 15 19 _ Vulkan _ 0 0 9-10164 _ God of War® III  BCUS98111

**Comment by elad335** [comment]
Try with tsx disabled.

**Comment by Andryshik345** [comment]
Tried some games on `AMD Ryzen 5 1600 3.4 GHz` and for me there's not so much degradation, maybe 1-3 fps.

<details>
  <summary>NFS ProStreet</summary>

With PR:
nfs with neko pr
Without PR:
nfs without neko pr

</details>
<details>
  <summary>Portal 2</summary>

With PR:
portal 2 with neko pr

Without PR:
portal 2 without neko pr

</details>

**Comment by psennermann** [comment]
Sorry for the ignorance, but is there a way to download this build? I mean, in the checks section I can't find a link to download it anymore...

**Comment by Nekotekina** [comment]
You need to download artifact from https://github.com/RPCS3/rpcs3/pull/8008/checks?check_run_id=582584162 - go to azure pipelines.

**Comment by psennermann** [comment]
With Ni no Kuni there doesn't seem to be any real difference (frames are not capped and same value is pure coincidence)

PR
Desktop Screenshot 2020 04 13 - 16 56 32 12


MASTER
Desktop Screenshot 2020 04 13 - 16 52 58 67

**Comment by Nekotekina** [comment]
Fixed a bug, need retest.

**Comment by dio-gh** [comment]
Tested with Persona 5 and Nier on my i7 7700HQ and to my absolute surprise, I saw no perf degradation. LGTM.

**Comment by lainiwaku** [comment]
tested on resistance 1/3  no difference

**Comment by Andryshik345** [comment]
No difference for me too, i think.
<details>
<summary>NFS ProStreet</summary>

with PR:
nfs with neko pr_2

master:
nfs without neko pr_2

</details>

**Comment by psennermann** [comment]
Heavenly Sword seems to have the same performance as well

**Comment by frymezim** [comment]
Tested it as well on my Intel 4770HQ laptop with Ni no kuni. The FPS is ranging from 15-26 on this build and same with master build.

**Pull request merged**